### ISSUE TITLE
HIGH: [Reliability] "Auth DeletePolicy" error occurs when deleting a workspace connected to a room

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#20860]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Log in to user1 account
2. Go to Settings > Workspaces and create a new workspace
4. Invite user2 to join the newly created workspace
4. Log in to user2 account
5. Navigate to the New Room section
6. Create a room and select workspace created by user1 and select room's visibility to public
7. Switch back to user1 account
8. Attempt to delete the workspace
## Expected Result: 
Auth error should not appear while deleting workspace
## Actual Result: 
"Auth DeletePolicy" error occurs when deleting a workspace connected to a room

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.28-3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/34695a5d-b055-441b-a17f-d2ed65180d7e

https://github.com/Expensify/App/assets/43996225/d78fc125-51a4-41c5-85f7-f203926955e8

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ayazhussain79 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686859691741469

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e90bc248e9218d8a</li>
        <li>Upwork Job ID: 1669695619332808704</li>
        <li>Last Price Increase: 2023-07-07</li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
Reproduced. I also noted this doesn't happen with other room types (like restricted).

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

"Auth DeletePolicy" error occurs when deleting a workspace connected to a room

### What is the root cause of that problem?

When the visibility is set to `private` or `public` then, `policyOwner` is not added as a room participant. When we choose `Restricted` visibility, all workspace members are included.

This will cause room not found when we perform [`DeleteWorkspace`](https://github.com/Expensify/App/blob/83ab0160a3b5c385fd26fffe20bdbccabf9249c0/src/libs/actions/Policy.js#L84) API request.

### What changes do you think we should make in order to solve the problem?

Ideally, if we execute `DeleteWorkspace` API by the `policyOwner`, it should find the room id even if the policy owner is not added to the room as a participant. 

### What alternative solutions did you explore? (Optional)
We can add all members to the room for all of the visibility types like what we did for the `Restrict` visibility type, or at least we can add the policy owner to the new room as a participant.

We should also modify the backend for this option, and then add `ownerAccountID` to participants, for optimistic report creation.

From this:
https://github.com/Expensify/App/blob/83ab0160a3b5c385fd26fffe20bdbccabf9249c0/src/libs/actions/Report.js#L1156-L1159

To this:
```
function addPolicyReport(policy, reportName, visibility) {
    const ownerAccountID = PersonalDetailsUtils.getAccountIDsByLogins([policy.owner])[0];
    // The participants include the current user (admin) and the employees. Participants must not be empty.
    const participants = _.unique([ownerAccountID, currentUserAccountID, ..._.pluck(policy.employeeList, 'accountID')]);
```

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by wildan-m** [comment]
@aimane-chnaif Yes, I can still reproduce it using 7c04c2f3f10df403f97fecb3b18dd81478003a31 version.

<img width="893" alt="image" src="https://github.com/Expensify/App/assets/11847279/99b90dd6-37d8-491d-8f5b-fb8c1b4bf94c">


Make sure:
1. The room creator is not the workspace owner
2. The visibility is `Public` or `Public`
3. The room creator has not accessed the room before clicking the `Delete Workspace` button

**Comment by joekaufmanexpensify** [comment]
I was also able to reproduce it as well again today.

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif Thoughts on the above proposal?

**Comment by joekaufmanexpensify** [comment]
Great, thanks!

**Comment by ShogunFire** [comment]
The error seems to happen in the backend when calling DeleteWorkspace
![image](https://github.com/Expensify/App/assets/19537677/1bcfc85b-f57d-4942-ae7e-8e4bf9a48ba5)
Offline it acts as intended.

**Comment by joekaufmanexpensify** [comment]
Still pending review of proposal

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif could you take a look at [this proposal](https://github.com/Expensify/App/issues/20860#issuecomment-1595601301) when you have a chance? Thanks!

**Comment by joekaufmanexpensify** [comment]
Still pending review of proposal

**Comment by joekaufmanexpensify** [comment]
Sure, happy to bump the question. Just to clarify though, do you mean that this issue should be on hold until the issue you referenced is complete, or just until the question in the SO is answered?

**Comment by joekaufmanexpensify** [comment]
Sounds good! I will look into getting an answer here.

**Comment by joekaufmanexpensify** [comment]
Still looking into this.

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif I saw the SO from above got answered. Curious if you have any more thoughts on whether this should now be held on the above issue?

**Comment by joekaufmanexpensify** [comment]
Ah, sure!

Question:
>We have three types of Workspace rooms: Public, Private and Restricted. What is the expected behavior for room visibility and when a member is added to a public, restricted or private room?

Answer: 
>Public: Anyone with the room URL can access the room
Restricted: Can be discovered and commented in by users that are members of the Workspace the room is tied to.
Private: Can only be discovered and commented in by users that are explicitly invited to the room.

**Comment by joekaufmanexpensify** [comment]
Ah, I see. I don't think joining the workspace causes the member to be added to any rooms automatically(beyond #announce). But I am clarifying!

**Comment by joekaufmanexpensify** [comment]
Stil discussing in Slack

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif I know we were discussing your question in [this Slack thread. ](https://expensify.slack.com/archives/C01GTK53T8Q/p1688761810432189?thread_ts=1687019472.887419&cid=C01GTK53T8Q) Is there any other info you need to determine whether this issue should now be on hold?

**Comment by mallenexpensify** [comment]
I was going to put this on hold pending 
- https://github.com/Expensify/App/issues/22699

And [this long discussion ](https://expensify.slack.com/archives/C01GTK53T8Q/p1688646398596839)
Sound good?

**Comment by joekaufmanexpensify** [comment]
I also was not thinking this relates to dynamically updating the app. Commented in the slack thread to get aligned on this.

**Comment by joekaufmanexpensify** [comment]
Seems like others agree with continuing on with this issue.

**Comment by joekaufmanexpensify** [comment]
>If so, when admin tries delete workspace which has the room created by employee but admin not joined yet, still allow deleting all the rooms regardless?

@aimane-chnaif I'm going to look into this question. Are there any others you need before we can decide how to proceed here?

**Comment by joekaufmanexpensify** [comment]
Sounds good!

**Comment by joekaufmanexpensify** [comment]
Asked [here.](https://expensify.slack.com/archives/C01GTK53T8Q/p1689273829871079?thread_ts=1687019472.887419&cid=C01GTK53T8Q)

**Comment by joekaufmanexpensify** [comment]
Got confirmation in Slack that if an admin deletes a workspace, we'll archive all of the workspace's rooms, even if the admin is not a member of all of them.

**Comment by joekaufmanexpensify** [comment]
@aimane-chnaif now that we have confirmation on your question, curious if we can move forward with fixing this one, or if it should be on hold? 

If we can fix it now, I'm going to immediately float it to one of the agencies to see if they can work on it.

**Comment by joekaufmanexpensify** [comment]
Not overdue

**Comment by joekaufmanexpensify** [comment]
Sounds good! Assigning the engineering label since it sounds like this needs to be internal. Here's a summary of the (many) comments on the issue: 

- Originally, we [were thinking](https://github.com/Expensify/App/issues/20860#issuecomment-1612008105) this should be on hold for another issue. The result of that discussion was that this does not need to be held on any other issues.
- Then, there was some separate discussion about whether this is related to other [dynamically updating issues](https://github.com/Expensify/App/issues/20860#issuecomment-1631541672), and whether we should fix this at all. The [result](https://expensify.slack.com/archives/C01GTK53T8Q/p1689190931091469?thread_ts=1688646398.596839&cid=C01GTK53T8Q) of that discussion was that this is not related, and we can go ahead.

**Comment by joekaufmanexpensify** [comment]
Pending confirmation from Georgia that this should be internal!

**Comment by joekaufmanexpensify** [comment]
@grgia Mind taking a look at this when you have a sec?

**Comment by grgia** [comment]
@joekaufmanexpensify or @aimane-chnaif  could you please summarize the findings/problem so far?

**Comment by joekaufmanexpensify** [comment]
Sure! There's a TLDR of the comments on the issue so far [here. ](https://github.com/Expensify/App/issues/20860#issuecomment-1638756844) Beyond that, the bug is basically that if user 1 creates a workspace, and invites user 2, and then user 2 creates a public room on the workspace, then user 1 cannot delete the workspace. After our investigation @aimane-chnaif thinks this needs to be internal. 

LMK if you need any more info beyond that!

**Comment by joekaufmanexpensify** [comment]
Not overdue

**Comment by joekaufmanexpensify** [comment]
Bumped 1:1

**Comment by joekaufmanexpensify** [comment]
@grgia bump when you have a chance, is there any more info you need beyond [this?](https://github.com/Expensify/App/issues/20860#issuecomment-1642374285)

**Comment by grgia** [comment]
Thank you for the bump, I'll take a look to determine

**Comment by grgia** [comment]
I can't get past step 3. inviting a user to the workspace on dev main, is that currently broken? @aimane-chnaif

**Comment by grgia** [comment]
Logging out and logging in fixed it

**Comment by grgia** [comment]
hmmm https://github.com/Expensify/Web-Expensify/blob/main/lib/ReportStore.php#L507-L508 this is the UUID error I'm seeing, which makes no sense as I'm coming from a web device.

**Comment by grgia** [comment]
I looked at the stack trace and the thrown report not found error is tied to the public room report


[info] [trace] StackTrace start
[info] [trace] 0/4. file ['/vagrant/Web-Expensify/lib/ReportStore.php' function 'load' line '68'](https://github.com/Expensify/Web-Expensify/blob/72f1a91b4f6fbce080d8bd1fa7cf0812b4f519bd/lib/ReportStore.php#L68)
[info] [trace] 1/4. file ['/vagrant/Web-Expensify/lib/ReportStore.php' function 'getByIDs' line '128'](https://github.com/Expensify/Web-Expensify/blob/72f1a91b4f6fbce080d8bd1fa7cf0812b4f519bd/lib/ReportStore.php#L128)
 [info] [trace] 2/4. file ['/vagrant/Web-Expensify/lib/ReportAPI.php' function 'getByID' line '4606'](https://github.com/Expensify/Web-Expensify/blob/main/lib/ReportAPI.php#L4606)
[info] [trace] 3/4. file ['/vagrant/Web-Expensify/lib/PolicyAPI.php' function 'readOldestAction' line '4810'](https://github.com/Expensify/Web-Expensify/blob/main/lib/PolicyAPI.php#L4810)
[info] [trace] 4/4. file ['/vagrant/Web-Expensify/api.php' function 'deleteWorkspace' line '2425'](https://github.com/Expensify/Web-Expensify/blob/main/api.php#L2425)

**Comment by joekaufmanexpensify** [comment]
Ah, interesting. That means we know exactly where/why we threw an error/exception now. Is that right?

**Comment by joekaufmanexpensify** [comment]
Opened a thread about this [here.](https://expensify.slack.com/archives/C02NK2DQWUX/p1690922316951579)

**Comment by grgia** [comment]
Going to try and work on this before EOW, if I can't, I'll post a summary of the next steps and look for a volunteer since I'm OOO next week.

**Comment by joekaufmanexpensify** [comment]
Sounds great, thanks Georgia!

**Comment by grgia** [comment]
Tested again on staging- logs https://staging.expensify.com/_devportal/tools/logSearch/#query=request_id:(%227f1035d80f4158d2-IAD%22)+AND+timestamp:[2023-08-03T16:16:46.838Z+TO+2023-08-03T18:16:46.838Z]&index=logs_expensify-023628

**Comment by grgia** [comment]
Bumping to weekly while I'm OOO.

**Comment by joekaufmanexpensify** [comment]
Got it, sg!

**Comment by joekaufmanexpensify** [comment]
@grgia if you're back this week, should we bump this one back up to daily?

**Comment by joekaufmanexpensify** [comment]
@grgia @aimane-chnaif is the next step here still an internal PR?

**Comment by joekaufmanexpensify** [comment]
[Bumped in Slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1693231752004189)

**Comment by joekaufmanexpensify** [comment]
Still pending internal PR

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by joekaufmanexpensify** [comment]
Pending internal PR

**Comment by grgia** [comment]
Hey @jasperhuangg this is an issue where after a member (NOT CREATOR) of a workspace creates a PUBLIC room, when the creator of the workspace tries to delete the workspace, there is a DeletePolicy error. Since you've been working on a lot. of the rooms projects, do you have any context to where this might be happening or who worked on public rooms logic and could take on this bug?

**Comment by joekaufmanexpensify** [comment]
@jasperhuangg bumping this for when you have a sec!

**Comment by jasperhuangg** [comment]
Hey, sorry for the delay! My GH notification settings were set up incorrectly.

I can assign myself to this and take a look.

**Comment by joekaufmanexpensify** [comment]
Great, ty!

**Comment by jasperhuangg** [comment]
Pretty swamped this week with App Deployer duties and some post-offshore
cleanup, but this is on my radar.

**Comment by joekaufmanexpensify** [comment]
TY!

**Comment by jasperhuangg** [comment]
The bug is coming from this line: https://github.com/Expensify/Web-Expensify/blob/main/lib/PolicyAPI.php#L5091

It seems the caller of `DeleteWorkspace` doesn't have the correct permissions for the public room that was created (since they haven't opened it before), so when they try to load it from Auth is gives us this permissions error.

I think it should be enough to wrap the call in a `try/catch` and do nothing if we get this error. With the new changes to how rooms work, it's actually very likely that a workspace admin/owner trying to delete a workspace won't have access to an underlying room.

PR incoming

**Comment by jasperhuangg** [comment]
The bug is coming from this line: https://github.com/Expensify/Web-Expensify/blob/main/lib/PolicyAPI.php#L5091

It seems the caller of `DeleteWorkspace` doesn't have the correct permissions for the public room that was created (since they haven't opened it before), so when they try to load it from Auth is gives us this permissions error.

This is actually a pretty weird problem. Here's an example of how it can happen:

1. userA creates workspace, invites userB
2. userB creates a room on the workspace, and invites userC. userA does not have access to this room.
3. userA deletes the workspace.
4. When userA is trying to delete the workspace, we archive the room created in [2], and marks actions sent in that room as read for userB and userC [here](https://github.com/Expensify/Web-Expensify/blob/main/lib/PolicyAPI.php#L5091).
5. However, since userA does not have access to the room, trying to mark the last action as read throws an exception.

It seems we need to move the logic that marks the latest action as read [here](https://github.com/Expensify/Web-Expensify/blob/9ea3085bb531f97faac8b842fcedbae20aae68f6/lib/PolicyAPI.php#L5089-L5091) to Auth.

**Comment by jasperhuangg** [comment]
I'm still pretty swamped with stuff for Money2020, but I can get a fix up for this once my plate clears up a bit.

**Comment by joekaufmanexpensify** [comment]
Sounds good. Thanks for the update Jasper!

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by jasperhuangg** [comment]
Still pretty swamped now that I'm trying to clear off my plate. If this needs to get done soon, maybe recruit another engineer to fix this?

I think the outline I've provided [here](https://github.com/Expensify/App/issues/20860#issuecomment-1765297826) should give more than enough pointers on how to solve this.

I'd say that this is even worthy of a `FirstPick` label.

**Comment by jasperhuangg** [comment]
@grgia do you have any bandwidth to take this over?

**Comment by jasperhuangg** [comment]
I'm going to unassign myself for now, but if no one has been able to prioritize this and my plate is looking better, feel free to reach out.

**Comment by joekaufmanexpensify** [comment]
got it, okay, sounds good!

**Comment by joekaufmanexpensify** [comment]
pending internal prioritization

**Comment by joekaufmanexpensify** [comment]
Same.

**Comment by joekaufmanexpensify** [comment]
Same

**Comment by joekaufmanexpensify** [comment]
No update

**Comment by joekaufmanexpensify** [comment]
I retested this one, and there's still an error deleting a workspace connected to a public room. @grgia curious if you think you'll be able to work on this one soon? This probably falls into the VIP-VSB bucket.

**Comment by joekaufmanexpensify** [comment]
We [decided](https://expensify.slack.com/archives/C01GTK53T8Q/p1705437972472019?thread_ts=1705437246.605389&cid=C01GTK53T8Q) to add this bug to VIP-VSB. So went ahead and did that!

**Comment by quinthar** [comment]
Hm, well I think this is actually touching on a bigger issue: it really doesn't make sense to delete workspaces when we don't delete the reports/rooms in that workspace.  We just "orphan" them, which makes everything strange.  I think the real solution is to actually *stop deleting workspaces*.

**Comment by joekaufmanexpensify** [comment]
Ah, that's interesting. Makes sense to me. I saw [your proposal](https://expensify.slack.com/archives/C066HJM2CAZ/p1705892706866269) for this, and I like it.

**Comment by joekaufmanexpensify** [comment]
Still pending internal prioritization

**Comment by quinthar** [comment]
Escalating to HIGH so we can get more discussion on this.

**Comment by quinthar** [comment]
Asking more about it here: https://expensify.slack.com/archives/C03U7DCU4/p1708850855478309

**Comment by joekaufmanexpensify** [comment]
Great. TY!

**Comment by joekaufmanexpensify** [comment]
Going to work on seeing if this issue is reproducible!

**Comment by joekaufmanexpensify** [comment]
Just retested this, and it's working now! If you create a public room on a workspace and then delete the workspace, you can now successfully delete it, and the public room is just archived.

**Comment by joekaufmanexpensify** [comment]
Closing as the reported bug is no longer occurring!

**Comment by ayazhussain79** [comment]
@joekaufmanexpensify could you please check, this bug was reported Jun 16, 2023 before closing the bug reporting, so I'm eligible for bug reporting right?

**Comment by joekaufmanexpensify** [comment]
No, I don't think so. We didn't fix anything in this issue.
